Add Shibboleth with OIDC Plugin Documentation (#15112)
matrix-org/synapse: docs/openid.md

        display_name_template: "{{ user.preferred_username|capitalize }}"
```

> ### Shibboleth with OIDC Plugin

Reviewer: Please order these alphabetically.

Author: done

[Shibboleth](https://www.shibboleth.net/) is an open Standard IdP solution widely used by Universities.

1. Shibboleth needs the [OIDC Plugin](https://shibboleth.atlassian.net/wiki/spaces/IDPPLUGINS/pages/1376878976/OIDC+OP) installed and working correctly.
2. Create a new config for the new endpoint on IDP Side:

> ```yaml
> {
>         "client_id": "SOME-CLIENT-ID",
>         "client_secret": "SOME-SUPER-SECRET-SECRET",
>         "response_types": ["code"],
>         "grant_types": ["authorization_code"],
>         "scope": "openid profile email",
>         "redirect_uris": ["https://[synapse public baseurl]/_synapse/client/oidc/callback"]
> }
> ```

Reviewer: ```suggestion
```json
{
    "client_id": "SOME-CLIENT-ID",
    "client_secret": "SOME-SUPER-SECRET-SECRET",
    "response_types": ["code"],
    "grant_types": ["authorization_code"],
    "scope": "openid profile email",
    "redirect_uris": ["https://[synapse public baseurl]/_synapse/client/oidc/callback"]
}
```
```

Synapse config:
```yaml
oidc_providers:
  # Shibboleth IDP
  #
  - idp_id: shibboleth
    idp_name: "Shibboleth Login"
    discover: true
    issuer: "https://YOUR-IDP-URL.TLD"

>     client_id: "YOUR_CLIENT_ID" #Note: not an URL because of Secial Chars

Reviewer: What does the comment here mean? "Not an URL"? Why would this be a URL?

Author: you maybe could also use a URL Like "https://myawesomeservice.org" as Client_ID but that seems to make problems here.

Reviewer: This should be something random, it looks like you have to fill this in manually as part of the OIDC plug-in?

Author: the client_id can be anything random, yes. But it must be the same on both sides: provider/client

Author: i did pull out the comment and did my hint below the code section.

Note that the fields `client_id` and `client_secret` are taken from the CURL response above.

### Shibboleth with OIDC Plugin

[Shibboleth](https://www.shibboleth.net/) is an open Standard IdP solution widely used by Universities.

1. Shibboleth needs the [OIDC Plugin](https://shibboleth.atlassian.net/wiki/spaces/IDPPLUGINS/pages/1376878976/OIDC+OP) installed and working correctly.

> 2. Create a new config for the new endpoint on IDP Side:

Reviewer: ```suggestion
2. Create a new config on the IdP Side, ensure that the `client_id` and `client_secret`
   are randomly generated data.
```

```json
{
    "client_id": "SOME-CLIENT-ID",
    "client_secret": "SOME-SUPER-SECRET-SECRET",
    "response_types": ["code"],
    "grant_types": ["authorization_code"],
    "scope": "openid profile email",
    "redirect_uris": ["https://[synapse public baseurl]/_synapse/client/oidc/callback"]
}
```

Synapse config:

```yaml
oidc_providers:
  # Shibboleth IDP
  #
  - idp_id: shibboleth
    idp_name: "Shibboleth Login"
    discover: true
    issuer: "https://YOUR-IDP-URL.TLD"
    client_id: "YOUR_CLIENT_ID"
    client_secret: "YOUR-CLIENT-SECRECT-FROM-YOUR-IDP"
    scopes: ["openid", "profile", "email"]
    allow_existing_users: true
    user_profile_method: "userinfo_endpoint"
    user_mapping_provider:
      config:
        subject_claim: "sub"
        localpart_template: "{{ user.sub.split('@')[0] }}"
        display_name_template: "{{ user.name }}"
        email_template: "{{ user.email }}"
```

> Hint: In my case using a URL like https://myawesomeservice.org as Client_ID lead to an error, so maybe better use something else with fewer special chars.
> 

Reviewer: You really don't want to use something like your URL as a client ID and it should be randomly generated. I've noted this above.

```suggestion
```
